akka-stream/src/main/scala/akka/stream/impl/Sources.scala

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import akka.stream.OverflowStrategies._
+import akka.stream._
+import akka.stream.stage._
+import scala.concurrent.{ Future, Promise }
+
+/**
+ * INTERNAL API
+ */
+private[akka] class QueueSource[T](maxBuffer: Int, overflowStrategy: OverflowStrategy) extends GraphStageWithMaterializedValue[SourceShape[T], SourceQueue[T]] {
+  trait SourceCompletion {
+    val completion = Promise[Unit]
+  }
+
+  type Offered = Promise[StreamCallbackStatus[Boolean]]
+
+  val out = Outlet[T]("queueSource.out")
+  override val shape: SourceShape[T] = SourceShape.of(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val buffer = if (maxBuffer == 0) null else FixedSizeBuffer[T](maxBuffer)
+    var pendingOffer: Option[(T, Offered)] = None
+    var pulled = false
+
+    val stageLogic = new GraphStageLogic(shape) with CallbackWrapper[(T, Offered)] with SourceCompletion {
+      def bufferOverflowException = new Fail.BufferOverflowException(s"Buffer overflow (max capacity was: $maxBuffer)!")
+
+      override def preStart(): Unit = initCallback(callback.invoke)
+      override def postStop(): Unit = stopCallback {
+        case (elem, promise) ⇒ promise.failure(new IllegalStateException("Stream is terminated. SourceQueue is detached"))
+      }
+
+      private def enqueueAndSuccess(elem: T, promise: Offered): Unit = {
+        buffer.enqueue(elem)
+        promise.success(StreamCallbackStatus.Success(true))
+      }
+
+      private def receiveElem(elem: T, promise: Offered): Unit = {
```

Comment:
please rename to `bufferElem` to make it clear that this is only for buffering cases